src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,49 +1,101 @@
 package core.basesyntax;
 
+import java.util.NoSuchElementException;
+
 public class ArrayList<T> implements List<T> {
+    private static final int DEFAULT_CAPACITY = 10;
+    private int size;
+    private int capacity = DEFAULT_CAPACITY;
+    private Object[] items = new Object[DEFAULT_CAPACITY];
 
     @Override
     public void add(T value) {
-
+        size++;
+        ensureCapacity();
+        items[size - 1] = value;
     }
 
     @Override
     public void add(T value, int index) {
-
+        size++;
+        if (isValidIndex(index)) {
+            ensureCapacity();
+            System.arraycopy(items, index, items, index + 1, size - index - 1);
+            items[index] = value;
+        } else {
+            throw new ArrayIndexOutOfBoundsException();
+        }
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        for (int i = 0; i < list.size(); i++) {
+            add(list.get(i));
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        if (isValidIndex(index)) {
+            return (T) items[index];
+        }
+        throw new ArrayIndexOutOfBoundsException();
     }
 
     @Override
     public void set(T value, int index) {
-
+        if (isValidIndex(index)) {
+            items[index] = value;
+        } else {
+            throw new ArrayIndexOutOfBoundsException();
+        }
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        if (isValidIndex(index)) {
+            final T temp = get(index);
+            size--;
+            System.arraycopy(items, index + 1, items, index, size - index);
+            items[size] = null;
+            return temp;
+        }
+        throw new ArrayIndexOutOfBoundsException();
     }
 
     @Override
     public T remove(T t) {
-        return null;
+        if (!isEmpty()) {
+            for (int i = 0; i < size; i++) {
+                if (t == null && get(i) == null || get(i).equals(t)) {
```

Comment:
what if `items[i]` is null and `t`  is not null?
```suggestion
                if (t == null && items[i] == null || items[i].equals(t)) {
```